forc-plugins/forc-doc/src/render.rs

```diff
@@ -300,6 +156,186 @@ impl RenderedDocumentation {
+                        doc_links.insert(BlockTitle::Enums, vec![doc.link()]);
+                    }
+                },
+                TraitDecl { .. } => match doc_links.get_mut(&BlockTitle::Traits) {
+                    Some(links) => links.push(doc.link()),
+                    None => {
+                        doc_links.insert(BlockTitle::Traits, vec![doc.link()]);
+                    }
+                },
+                AbiDecl { .. } => match doc_links.get_mut(&BlockTitle::Abi) {
+                    Some(links) => links.push(doc.link()),
+                    None => {
+                        doc_links.insert(BlockTitle::Abi, vec![doc.link()]);
+                    }
+                },
+                StorageDecl { .. } => match doc_links.get_mut(&BlockTitle::ContractStorage) {
+                    Some(links) => links.push(doc.link()),
+                    None => {
+                        doc_links.insert(BlockTitle::ContractStorage, vec![doc.link()]);
+                    }
+                },
+                FunctionDecl { .. } => match doc_links.get_mut(&BlockTitle::Functions) {
+                    Some(links) => links.push(doc.link()),
+                    None => {
+                        doc_links.insert(BlockTitle::Functions, vec![doc.link()]);
+                    }
+                },
+                ConstantDecl { .. } => match doc_links.get_mut(&BlockTitle::Constants) {
+                    Some(links) => links.push(doc.link()),
+                    None => {
+                        doc_links.insert(BlockTitle::Constants, vec![doc.link()]);
+                    }
+                },
+                _ => {} // TODO: ImplTraitDeclaration
+            }
+        }
+        None => {
+            let mut doc_links: BTreeMap<BlockTitle, Vec<DocLink>> = BTreeMap::new();
+            match doc.item_body.ty_decl {
+                StructDecl { .. } => {
+                    doc_links.insert(BlockTitle::Structs, vec![doc.link()]);
+                }
+                EnumDecl { .. } => {
+                    doc_links.insert(BlockTitle::Enums, vec![doc.link()]);
+                }
+                TraitDecl { .. } => {
+                    doc_links.insert(BlockTitle::Traits, vec![doc.link()]);
+                }
+                AbiDecl { .. } => {
+                    doc_links.insert(BlockTitle::Abi, vec![doc.link()]);
+                }
+                StorageDecl { .. } => {
+                    doc_links.insert(BlockTitle::ContractStorage, vec![doc.link()]);
+                }
+                FunctionDecl { .. } => {
+                    doc_links.insert(BlockTitle::Functions, vec![doc.link()]);
+                }
+                ConstantDecl { .. } => {
+                    doc_links.insert(BlockTitle::Constants, vec![doc.link()]);
+                }
+                _ => {} // TODO: ImplTraitDeclaration
+            }
+            module_map.insert(module_prefixes.clone(), doc_links);
+        }
+    }
+}
```

Comment:
I think this can be shortened using a helper method:

```
fn populate_doc_links_for_decl(doc: &Document, doc_links: &mut BTreeMap<BlockTitle, Vec<DocLink>>) {
    let key = doc.item_body.ty_decl.as_block_title();
    match doc_links.get_mut(&key) {
        Some(links) => links.push(doc.link()),
        None => {
            doc_links.insert(key, vec![doc.link()]);
        }
    }
}

fn populate_doc_links(
    doc: &Document,
    module_map: &mut BTreeMap<ModulePrefixes, BTreeMap<BlockTitle, Vec<DocLink>>>,
) {
    let module_prefixes = &doc.module_info.module_prefixes;
    let doc_links = module_map
        .get_mut(module_prefixes)
        .unwrap_or(&mut BTreeMap::<BlockTitle, Vec<DocLink>>::new());
    populate_doc_links_for_decl(doc, doc_links);
    module_map.insert(module_prefixes.clone(), doc_links);
}
```